src/Parallafka/CommitPoller.cs

```diff
@@ -0,0 +1,162 @@
+        {
+            this._committer = committer;
+            this._cancellation = new CancellationTokenSource();
+            this._cancellationToken = _cancellation.Token;
+            this._state = "starting";
+
+            this.Completion = CommitLoop();
+        }
+
+        public object GetStats()
+        {
+            return new
+            {
+                State = this._state,
+                TotalCommits = this._totalCommits,
+                DelayedCommits = this._commitDelays,
+                QueueFullCommits = this._commitNows
+            };
+        }
+
+        public void Complete()
+        {
+            lock (this._taskLock)
+            {
+                this._complete = true;
+            }
+
+            this._cancellation.Cancel();
+        }
+
+        public Task Completion { get; }
+
+        /// <summary>
+        /// Returns true if the commit was queued, false if it was already queued or the poller is Complete
+        /// </summary>
+        public bool CommitWithin(TimeSpan delay)
+        {
+            lock (this._taskLock)
+            {
+                if (!this._complete)
+                {
+                    return this._commitDelay.TrySetResult(delay);
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Returns true if the commit was queued, false if it was already queued or the poller is Complete
+        /// </summary>
+        public bool CommitNow()
+        {
+            lock (this._taskLock)
+            {
+                if (!this._complete)
+                {
+                    return this._commitNow.TrySetResult();
+                }
+            }
+
+            return false;
+        }
+
+        private readonly object _taskLock = new();
+        private TaskCompletionSource _commitNow = new();
+        private TaskCompletionSource<TimeSpan> _commitDelay = new();
+        private long _totalCommits;
+        private long _commitDelays;
+        private long _commitNows;
+
+        private async Task CommitLoop()
+        {
+            while (true)
+            {
+                this._state = "idle";
+
+                var delayTask = Task.Delay(-1, this._cancellationToken);
+                var result = await Task.WhenAny(
+                    delayTask,
+                    this._commitNow.Task,
+                    this._commitDelay.Task);
+
+                if (result == delayTask)
+                {
+                    // see if there's any work to be done before loop completion
+                    if (this._commitNow.Task.IsCompletedSuccessfully || this._commitDelay.Task.IsCompletedSuccessfully)
+                    {
+                        await PerformCommit();
+                    }
+
+                    break;
+                }
+                
+                if (result == this._commitNow.Task)
+                {
+                    Interlocked.Increment(ref this._commitNows);
```

Comment:
Are these interlocked instead of making the members volatile for reads? Looks like they're only updated serially

Reply:
Are you asking why I just don't use `+=` inside the lock statement?  Mostly just to keep the lock statement as small as possible.  Interlocked.Increment is a single CPU-based operation and so is 100% efficient in its own right.  I shy away from volatile read/writes since I think those generally hide away/complicate multi-threading issues.  I'll prefer to use various Interlocked methods instead.

Reply:
Wondering why it's not just `+=` _outside_ the lock statement - I'm not seeing where it could update those in parallel so I was wondering if I'm missing how that happens or if you did it with Interlocked just to ensure the reading threads (GetStats callers) see updates ASAP.

Reply:
I wasn't worried about reader threads, only writer threads.  But you're right, only this running thread (task) is updating the value at any given time.  `++` would work as effectively as Interlocked.  I still prefer Interlocked just out of habit for things like this, I suppose.

Reply:
Fair, just wanted to make sure I wasn't missing some concurrency